### DISCUSSION TITLE
can i allow to edit records in tables only for record owners?

### DISCUSSION THREAD

**Discussion by vetsinen** [piccolo-orm/piccolo_admin#250] [General]
can i allow to edit records in tables only for record owners, if owners id is saved as foreign key?

**Comment by sinisaos** [comment]
@vetsinen You can try using the new Piccolo Admin feature and write custom [validators](https://piccolo-admin.readthedocs.io/en/latest/table_config/index.html#validators) for this purpose and pass them to TableConfig. Of course, the superuser can still do anything on the tables. Hope that helps.

**Comment by dantownsend** [comment]
Yeah, as @sinisaos said - validators are probably the best approach.

We just added support for async validators in Piccolo API - so make sure you upgrade to the latest version.

You can then do something like this:

```python
from piccolo_api.crud.endpoints import PiccoloCRUD
from piccolo_api.session_auth.tables import SessionsBase as _SessionsBase
from piccolo.columns.column_types import Varchar, ForeignKey
from piccolo.apps.user.tables import BaseUser as _BaseUser
from piccolo.table import Table, create_db_tables_sync
from piccolo_admin.endpoints import create_admin, TableConfig
from piccolo_api.crud.validators import Validators
from piccolo.engine.sqlite import SQLiteEngine
from starlette.exceptions import HTTPException
from starlette.requests import Request


DB = SQLiteEngine()


class BaseUser(_BaseUser, db=DB):
    pass


class SessionsBase(_SessionsBase, db=DB):
    pass


class MyTable(Table, db=DB):
    data = Varchar()
    user = ForeignKey(BaseUser, null=False)


###############################
# The important bit

async def owner_only(piccolo_crud: PiccoloCRUD, request: Request):
    row_id = request.path_params["row_id"]

    if not await MyTable.exists().where(
        MyTable.id == int(row_id),
        MyTable.user == request.user.user
    ):
        raise HTTPException('Only the owner can edit this row.')


app = create_admin(
    tables=[
        BaseUser,
        TableConfig(MyTable, validators=Validators(patch_single=[owner_only])),
    ],
    auth_table=BaseUser,
    session_table=SessionsBase,
)

###############################

if __name__ == '__main__':
    create_db_tables_sync(MyTable, BaseUser, SessionsBase, if_not_exists=True)

    if not BaseUser.exists().where(BaseUser.username == 'piccolo').run_sync():
        BaseUser.create_user_sync(
            username='piccolo',
            password='piccolo123',
            email="piccolo@test.com",
            admin=True,
            active=True,
            superuser=True
        )

    if not BaseUser.exists().where(BaseUser.username == 'other_user').run_sync():
        BaseUser.create_user_sync(
            username='other_user',
            password='other_user123',
            email="other_user@test.com",
            admin=True,
            active=True,
            superuser=True
        )

    import uvicorn
    uvicorn.run(app)
```

**Reply by sinisaos** [reply]
@vetsinen Thanks to @dantownsend now with the latest Piccolo API version ([0.49.0](https://github.com/piccolo-orm/piccolo_api/releases/tag/0.49.0)) you can also use async validators. Here is example of the working code.
```python
# table.py
from piccolo.apps.user.tables import BaseUser
from piccolo.columns import Varchar, Boolean, ForeignKey
from piccolo.columns.readable import Readable
from piccolo.table import Table


class Task(Table):
    name = Varchar()
    completed = Boolean(default=False)
    owner = ForeignKey(references=BaseUser)

    @classmethod
    def get_readable(cls):
        return Readable(template="%s", columns=[cls.owner])
```
and then in app.py you can write something like this
```python
# all your imports here
async def owner_only(piccolo_crud: PiccoloCRUD, request: Request):
    user = request.user.user
    # remove these two lines even if you want the superuser to not be able to edit on other owners records
    if user.superuser: 
        return
    task_owner = await Task.exists().where(Task.owner == user["id"])
    if not task_owner:
        raise HTTPException(
            status_code=405,
            detail="Only task_owners and superuser are allowed to do this",
        )


task_config = TableConfig(
    table_class=Task,
    validators=Validators(patch_single=[owner_only]),
)

app = FastAPI(
    routes=[
        Route("/", HomeEndpoint),
        Mount(
            "/admin/",
            create_admin(
                tables=[task_config],
            ),
        ),
        Mount("/static/", StaticFiles(directory="static")),
    ],
)
```
@dantownsend Great work on async validators and I can confirm that everything works fine with the latest Piccolo API version.

**Reply by dantownsend** [reply]
@sinisaos Thanks. I think the async validators will be useful.

**Reply by sinisaos** [reply]
@dantownsend Could you please correct the two typos in the `TableConfig` docstring about usage of `validators`

https://github.com/piccolo-orm/piccolo_admin/blob/e646bb7a00d9cdd55857064b8660bc747f72aed9/piccolo_admin/endpoints.py#L139

It should be `(Managers.user == user["id"])`

https://github.com/piccolo-orm/piccolo_admin/blob/e646bb7a00d9cdd55857064b8660bc747f72aed9/piccolo_admin/endpoints.py#L152

It should be list of validators like this `validators=Validators(post_single=[manager_only])`.

Thanks.

**Reply by dantownsend** [reply]
Yes, you're right - it should be a list. I'll change it to `validators=Validators(post_single=[manager_only])`.

I think this should work `manager = await Manager.exists().where(manager.user == user)` - Piccolo should automatically convert a table instance into the id.

**Reply by sinisaos** [reply]
> I think this should work `manager = await Manager.exists().where(manager.user == user)` - Piccolo should automatically convert a table instance into the id.

Yes. You're right. Sorry about that.

**Reply by dantownsend** [reply]
No worries. I've created a PR to fix it: https://github.com/piccolo-orm/piccolo_admin/pull/253

Unfortunately a bunch of the tests are failing, so will have to fix them first. Starlette's test client now uses `httpx` instead of `requests` under the hood, which is making a lot of the tests fail in subtle ways.

**Reply by sinisaos** [reply]
Great. Thanks. I see you are having the same problems in the Piccolo API with this braking backwards compatibility with the new `Starlette` Testclient. I hope it's not too much work.

**Reply by dantownsend** [reply]
I'll try and fix the tests this evening. Starlette 0.21.0 has been such a huge pain in the ass! It broke the Piccolo integration tests multiple times too. But now all of the ASGI libraries support this new version, life will be much easier. And `httpx` is superior to `requests` so I can see why they did it.
